crates/gitbutler-branch-actions/src/reorder_commits.rs

```diff
@@ -0,0 +1,639 @@
+
+use anyhow::{bail, Context as _, Result};
+use gitbutler_branch::{signature, BranchId, SignaturePurpose};
+use gitbutler_cherry_pick::RepositoryExt as _;
+use gitbutler_command_context::CommandContext;
+use gitbutler_commit::commit_ext::CommitExt as _;
+use gitbutler_project::access::WorktreeWritePermission;
+use gitbutler_repo::{rebase::cherry_rebase_group, LogUntil, RepositoryExt as _};
+
+use crate::{branch_trees::checkout_branch_trees, VirtualBranchesExt as _};
+
+///
+/// Presume we had the stack:
+///
+/// A
+/// B
+/// C
+/// D
+///
+/// If C was the subject, and the offset was -1, we would expect:
+///
+/// A
+/// C
+/// B
+/// D
+///
+/// Presume we had the stack:
+///
+/// A
+/// B
+/// C
+/// D
+///
+/// If B was the subject, and the offset was 1, we would expect:
+///
+/// A
+/// C
+/// B
+/// D
+///
+
+// move a given commit in a branch up one or down one
+// if the offset is positive, move the commit down one
+// if the offset is negative, move the commit up one
+// rewrites the branch head to the new head commit
+pub(crate) fn reorder_commit(
+    ctx: &CommandContext,
+    branch_id: BranchId,
+    subject_commit_oid: git2::Oid,
+    offset: i32,
+    perm: &mut WorktreeWritePermission,
+) -> Result<()> {
+    let repository = ctx.repository();
+    let vb_state = ctx.project().virtual_branches();
+    let default_target = vb_state.get_default_target()?;
+    let default_target_commit = repository
+        .find_reference(&default_target.branch.to_string())?
+        .peel_to_commit()?;
+    let merge_base = repository.merge_base(default_target_commit.id(), subject_commit_oid)?;
+
+    let mut branch = vb_state.get_branch_in_workspace(branch_id)?;
+    let subject_commit = repository
+        .find_commit(subject_commit_oid)
+        .context("failed to find commit")?;
+
+    let ReorderResult { tree, head } = inner_reorder_commit(
+        repository,
+        &repository.find_commit(merge_base)?,
+        &subject_commit,
+        offset,
+        &repository
+            .log(branch.head, LogUntil::Commit(merge_base))?
```

Comment:
Are the actual `Commit` objects needed here, or would `Oid` also do the trick?

Reply:
No, I often start with the Commit objects because they're more strongly typed and it may reduce excessive lookups, but they're certainly not needed here

Reply:
I think it's also related to `log()` returning commits by default. What's hidden in `git2` will be obvious with `gix`, and this method (`log`) will definitely start returning `ids` or graph walk items (once backed by `gix`), which allow access to parent-ids and commit-times, as free by-product of the iteration.
Everything else would then be re-obtained on demand only which will still be close-to-free as the objects are usually cached.
If you see no future need for these `Commit` objects, or the need for information on `Commit` only for a few select ones, then it would be beneficial to not use `Commit` objects here in preparation for the future.

Reply:
We have the horrendously named `l` function which returns just the `Oid`s for us.

Btw, when I said `No, ...` I meant, "No, they're no longer needed", and I've updated the functions to only pass the objects between themselves if they're actually needed.

It would be nice if we could have typed Oids because it can be clear looking at a function signature to tell what kind it wants (most of time time its a commit Oid, but our `branch.tree` property is actually the Oid of a tree 😆.

Reply:
> It would be nice if we could have typed Oids because it can be clear looking at a function signature to tell what kind it wants (most of time time its a commit Oid, but our `branch.tree` property is actually the Oid of a tree 😆.

I see, that would indeed be good to have. It would be somewhat expensive to do in `gix` as it would have to truly assure the object is of the right type which is a `find_header()` call at least - it's fast, but far from free.
Meanwhile, I think descriptive variable names can go a long way, also knowing that once one tries to get a tree with an Oid of a commit, it will fail as loud as one wishes.